horovod/common/operations.cc

```diff
@@ -1001,51 +1001,58 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
                               DDL_OP_SUM))
 #else
       if (horovod_global.hierarchical_allreduce) {
-        NCCL_CHECK(entries, "ncclReduce",
-                   ncclReduce(fused_input_data, buffer_data,
-                              (size_t)num_elements,
-                              GetNCCLDataType(first_entry.tensor), ncclSum, 0,
+        int64_t num_elements_per_gpu = num_elements%horovod_global.local_size == 0 ?
+                   (num_elements/horovod_global.local_size) : (num_elements/horovod_global.local_size + 1);
+        size_t buffer_len_per_proc = (buffer_len/num_elements)*num_elements_per_gpu;
+
+        NCCL_CHECK(entries, "ncclReduceScatter",
+                   ncclReduceScatter(fused_input_data, buffer_data,
```

Comment:
Does this work well with an in-place operation, where `fused_input_data = buffer_data` (tensor fusion use case)?

Reply:
I followed the instructions in https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/index.html#inplaceop to make `ncclAllGather` and `ncclReduceScatter` operate in-place.